Src/WindowsServer/WindowsServer.Shared/Implementation/AzureComputeMetadataHeartbeatPropertyProvider.cs

```diff
@@ -0,0 +1,114 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer.Implementation
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+
+    internal class AzureComputeMetadataHeartbeatPropertyProvider
+    {
+        internal const string HeartbeatPropertyPrefix = "azInst_"; // to ensure no collisions with base heartbeat properties
+
+        /// <summary>
+        /// Expected fields extracted from Azure IMS to add to the heartbeat properties. 
+        /// Set as internal for testing.
+        /// </summary>
+        internal readonly IReadOnlyCollection<string> ExpectedAzureImsFields = new string[]
+        {
+            "location",
+            "name",
+            "offer",
+            "osType",
+            "placementGroupId",
+            "platformFaultDomain",
+            "platformUpdateDomain",
+            "publisher",
+            "resourceGroupName",
+            "sku",
+            "subscriptionId",
+            "tags",
+            "version",
+            "vmId",
+            "vmSize"
+        };
+
+        /// <summary>
+        /// Flags that will tell us whether or not Azure VM metadata has been attempted to be gathered or not, and
+        /// if we should even attempt to look for it in the first place. 
+        /// </summary>
+        private bool isAzureMetadataCheckCompleted = false;
+
+        private IAzureMetadataRequestor azureInstanceMetadataRequestor = null;
```

Comment:
readonly ?

Reply:
BTW, you don't need to explicitly initialize to null (and `false` on line 40). All fields are guaranteed to be default-initialized. I thought there was a compiler warning about that?

Reply:
Not sure I understand what you mean here. The `azureInstanceMetadataRequestor` field could be set to readonly, and set in the constructor and not given this default value of null, but what would that achieve?

Reply:
Interesting, the 2nd comment you gave me here didn't show up as I was going through them. No compiler warning that I've seen, but we definitely have a few warnings to clean up in our code base. Still curious about the benefit of making the requestor field readonly.

Reply:
> Still curious about the benefit of making the requestor field readonly.

The general rule is: If the field (or property) is written to only in the constructor, then make it readonly. For private fields, it prevents unintended modification. It's an extra safety check. (In theory, it also opens up further optimizations but, in practice those are very rare).

Regarding the default-initialized fields: I know R# warns about them. I thought either the compiler, StyleCop or FxCop warned about them too. In any case, I think it's not generally accepted in the codebases I've worked on to do explicit default initialization and, indeed, it's frowned upon. This is the opposite of C++, but it is one of the design principals of .NET: eliminate memory errors due to garbage values.